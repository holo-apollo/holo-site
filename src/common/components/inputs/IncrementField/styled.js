// @flow
import styled from 'styled-components';

import palette from 'common/palette';

export const Cont = styled.div`
  display: flex;
  align-items: center;
`;

export const StyledInput = styled.input`
  min-width: 36px;
  height: 36px;
  border: solid 1px ${palette.darkGrey};
  font-size: 14px;
  color: ${palette.darkGrey};
  text-align: center;
  -webkit-appearance: none;
  font: inherit;
  outline: none;
`;

export const Modifier = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 29px;
  height: 30px;
  background-color: ${palette.lightestGrey};
  font-size: 24px;
  color: ${palette.darkGrey};
  cursor: pointer;
  user-select: none;

  &:hover {
    background: ${palette.lightestEvening};
  }
`;
