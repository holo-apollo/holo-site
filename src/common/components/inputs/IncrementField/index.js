// @flow
import React, { useState } from 'react';

import { Cont, StyledInput, Modifier } from './styled';

type Props = {
  initialValue: number,
  maxValue?: number,
  minValue?: number,
  step: number,
  onChange?: number => void,
  isPercent: boolean,
};

const defaultProps = {
  initialValue: 0,
  step: 1,
  isPercent: false,
};

const IncrementField = ({
  initialValue,
  minValue,
  maxValue,
  step,
  onChange,
  isPercent,
}: Props) => {
  const [value, setValue] = useState(initialValue);

  function decrement() {
    const nextValue = value - step;
    if (minValue === undefined || nextValue >= minValue) {
      setValue(nextValue);
      onChange && onChange(nextValue);
    }
  }

  function increment() {
    const nextValue = value + step;
    if (maxValue === undefined || nextValue <= maxValue) {
      setValue(nextValue);
      onChange && onChange(nextValue);
    }
  }

  const strValue = isPercent ? `${value}%` : value.toString();

  return (
    <Cont>
      <Modifier onClick={decrement}>-</Modifier>
      <StyledInput
        type="text"
        value={strValue}
        disabled={true}
        size={strValue.length}
      />
      <Modifier onClick={increment}>+</Modifier>
    </Cont>
  );
};

IncrementField.defaultProps = defaultProps;

export default IncrementField;
