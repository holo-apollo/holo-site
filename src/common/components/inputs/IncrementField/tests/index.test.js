import React from 'react';
import renderer from 'react-test-renderer';

import IncrementField from '../index';

describe('IncrementField', function() {
  it('should match snapshot', function() {
    const tree = renderer.create(<IncrementField />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
