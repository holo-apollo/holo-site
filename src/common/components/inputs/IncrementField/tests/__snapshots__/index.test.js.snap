// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncrementField should match snapshot 1`] = `
<div
  className="styled__Cont-sc-1uczn2-0 jiDxzY"
>
  <div
    className="styled__Modifier-sc-1uczn2-2 bLawfL"
    onClick={[Function]}
  >
    -
  </div>
  <input
    className="styled__StyledInput-sc-1uczn2-1 hjmKSG"
    disabled={true}
    size={1}
    type="text"
    value="0"
  />
  <div
    className="styled__Modifier-sc-1uczn2-2 bLawfL"
    onClick={[Function]}
  >
    +
  </div>
</div>
`;
